### ISSUE TITLE
Matrix multiplication of two LazyTensors

### ISSUE THREAD

**Issue by adam-coogan** [getkeops/keops#35]
Hello, thanks for the nice package!

I am running into a problem when trying to decompose a matrix as the product of two `LazyTensor`s. The matrix I want to represent has shape `[N, N]` and can be expressed as the product of an `[M, N]` matrix and its transpose. The snippet below shows what I'd like to do in more detail:
```python
import torch
from pykeops.torch import LazyTensor

# Set up inputs
M, N, d = 10, 5, 2
x, y = torch.rand([M, d]), torch.rand([N, d])

# Construct kernel matrix
x_i = LazyTensor(x[:, None, :])  # (M, 1, 2)
y_j = LazyTensor(y[None, :, :])  # (1, N, 2)
D_ij = ((x_i - y_j) ** 2).sum(-1)   # (M, N): squared distances
sqrt_K = (-D_ij).exp()

K = sqrt_K.t() @ sqrt_K  # does not work
# next step: run K.solve(...)
```
The last line does not work since `__matmul__` calls `view()` on its argument, which is not supported by `LazyTensor`. I also don't see how to construct a reduction formula for the matrix `K` since it seems like these can only involve two indices, while three are needed here. Is there some other way I can construct the matrix `K`?

**Comment by jeanfeydy** [comment]
Hi @adam-coogan ,

You're welcome!
Indeed, as you've remarked, the "matrix-matrix" product is not supported as a "symbolic" operation by KeOps LazyTensors. This is mainly a matter of timing: as a first step for the years 2018-2020, we decided to prioritize the support of low-level "pointwise" formulas and the development of multiple binders (Python, Matlab, R...), to the detriment of higher-level features. 
Long-term, we will extend the engine to support e.g. block-wise definitions for LazyTensors (which may be important for Finite/Boundary Elements simulations, ...), but don't expect these to come to soon.

Fortunately though, in most use cases (especially for linear algebra), you can *already* emulate most high-level behaviors using explicit linear operators. In your example, you could simply define a new "symbolic matrix" along the lines of:
```
K = lambda v : sqrt_K.t() @ (sqrt_K @ v)
```
and plug it into your favorite linear algebra solver.

The whole procedure is described in [this tutorial](https://www.kernel-operations.io/keops/_auto_tutorials/backends/plot_scipy.html), which relies on the very convenient interface of the [`scipy.sparse.linalg`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.linalg.LinearOperator.html) package. As discussed in [this benchmark](https://www.kernel-operations.io/keops/_auto_benchmarks/plot_benchmark_invkernel.html), off-the-shelf performances with the SciPy+KeOps combo are already very competitive for non-sparse systems. Doing better is essentially a research problem: you may want to have a look at our [GPytorch+KeOps tutorial](https://www.kernel-operations.io/keops/_auto_tutorials/backends/plot_gpytorch.html).

Best regards,

Jean

**Comment by adam-coogan** [comment]
Thanks for the reply! I should have been more clear in my original post, though. The matrix `sqrt_K` is dependent on some parameters, and I need to be able to differentiate the output of `K.solve(...)` with respect to these. This doesn't seem possible using the scipy solvers, and the gpytorch conjugate gradient solver also does not appear to be autograd-compatible. Do you have any other suggestions about how to work around this?

**Comment by jeanfeydy** [comment]
Hi @adam-coogan ,

Sorry for the late answer!
As far as I can tell, your best bet is to write an explicit [torch.autograd.Function](https://pytorch.org/tutorials/beginner/examples_autograd/two_layer_net_custom_function.html) for your custom solve operator. At heart, all differentiable solvers simply use the well-known formula for the [gradient of the inverse matrix](https://math.stackexchange.com/questions/1471825/derivative-of-the-inverse-of-a-matrix/1471836). Using the notations of [our autograd documentation](http://www.kernel-operations.io/keops/autodiff_gpus/backpropagation.html), if:

f : (K, y) -> K \ y = x

is your forward "linsolve" operator, then its gradient wrt. K reads:

\partial_K [ f ] . dx = - (K^T \ dx) @ x^T

(GitHub doesn't support LaTex :'-( )
(It's super-cryptic, but I think that this is actually what is implemented [here](https://github.com/getkeops/keops/blob/master/pykeops/torch/operations.py). Is it correct @joanglaunes ? I'm writing this up quickly, so don't take my word for the computations!)
Basically, backpropagating through a "linsolve" only requires you to solve another linear system with the transpose of K. In your case, you could simply derive the chain rule by hand for:

g : (K, y) ->  (K * K^T) \ y = x

which reads, if I'm not mistaken (10% likelihood...): 

\partial_K [ g ] . dx = - [ ((K@K^T) \ dx) @ (K^T * x)^T  +  x @ (K^T @ ( (K@K^T) \ dx) ) ]

and implement this formula implicitely, just like in the regular [KeOps linsolve](https://github.com/getkeops/keops/blob/master/pykeops/torch/operations.py).
Unfortunately, I'm not 100% sure of how to do it, but @joanglaunes could probably give us some insight here!

Best regards,
Jean

**Comment by joanglaunes** [comment]
Hello @adam-coogan ,

I just finished to implement this operation in the branch "squared_kernel"
[https://github.com/getkeops/keops/tree/squared_kernel](url)
I have named sqsolve the corresponding LazyTensor operation (I am not sure it is a good nickname..), see the example below.
As Jean was explaining, it involved writing a custom autograd function, which is quite cumbersome! In fact I am still hoping to find a way to be more generic, so that we could just apply "solve" to compositions of KeOps operations. This is why I am not merging this branch to master yet.

Here is the test program :
```python

# Example of use of the "sqsolve" operation on LazyTensors
# Given M random points x_i, N random points y_j, N random values b_j,
# and a scale parameter sigma,
# we solve the matrix system A^T A u = b
# where A is the M*N kernel matrix A_ij = exp(-|x_i-y_j|^2/sigma^2)
# We also test the automatic differentiation of this operation

import torch
from pykeops.torch import LazyTensor

# Set up inputs
M, N, d = 10, 5, 2
x, y, b = torch.rand([M, d],requires_grad=True), torch.rand([N, d],requires_grad=True), torch.randn([N, 1],requires_grad=True)
sigma = 0.25

# Construct kernel matrix
x_i = LazyTensor(x[:, None, :])  # (M, 1, 2)
y_j = LazyTensor(y[None, :, :])  # (1, N, 2)
b_j = LazyTensor(b[None, :, :])  # (1, N, 1)
D_ij = ((x_i - y_j) ** 2).sum(-1)   # (M, N): squared distances
sqrt_K = (-D_ij/sigma**2).exp()

res_keops = sqrt_K.sqsolve(b_j)
loss = (res_keops**2).sum()
loss.backward()

# same with Pytorch

xc, yc, bc = torch.rand([M, d],requires_grad=True), torch.rand([N, d],requires_grad=True), torch.randn([N, 1],requires_grad=True)
xc.data, yc.data, bc.data = x.data, y.data, b.data

x_i = xc[:, None, :]  # (M, 1, 2)
y_j = yc[None, :, :]  # (1, N, 2)
D_ij = ((x_i - y_j) ** 2).sum(-1)   # (M, N): squared distances
sqrt_K = (-D_ij/sigma**2).exp()
K = sqrt_K.t() @ sqrt_K

res_torch = torch.solve(bc,K)[0]
loss = (res_torch**2).sum()
loss.backward()

errf = (torch.norm(res_keops-res_torch)/torch.norm(res_torch)).item()
errgx = (torch.norm(x.grad-xc.grad)/torch.norm(xc.grad)).item()
errgy = (torch.norm(y.grad-yc.grad)/torch.norm(yc.grad)).item()
errgb = (torch.norm(b.grad-bc.grad)/torch.norm(bc.grad)).item()

print("relative error forward : ", errf)
print("relative error grad x : ", errgx)
print("relative error grad y : ", errgy)
print("relative error grad b : ", errgb)

```

Please let us know if you can test it and if it works on your side ; also if you found another way to do this since your last message.

Best,
Joan

**Comment by adam-coogan** [comment]
This works for me, thanks so much for implementing this!
